### PR TITLE
[spinel] handle link metrics

### PR DESCRIPTION
**Pull request by piotrkoziar**
This PR allows the spinel interface to deal with link metrics.

There are 7 new properties introduced:
* `THREAD_LINK_METRICS_QUERY`
* `THREAD_LINK_METRICS_QUERY_RESULT`
* `THREAD_LINK_METRICS_PROBE`
* `THREAD_LINK_METRICS_MGMT_ENH_ACK`
* `THREAD_LINK_METRICS_MGMT_ENH_ACK_IE`
* `THREAD_LINK_METRICS_MGMT_FORWARD`
* `THREAD_LINK_METRICS_MGMT_RESPONSE`

### CODE REVIEW

**Comment by konradderda** [comment]
@googlebot I consent

File: src/lib/spinel/spinel.h [hunk 1]
```diff
@@ -2944,6 +2984,160 @@ enum
...
+    SPINEL_PROP_THREAD_LINK_METRICS_MGMT_ENH_ACK_IE = SPINEL_PROP_THREAD_EXT__BEGIN + 49,
+
+    /// Link metrics Forward Tracking Series management
+    /** Format: 6Cdd - Write only
+     *
+     * Required capability: `SPINEL_CAP_THREAD_LINK_METRICS`
+     *
+     * `6` : IPv6 destination address
+     * `C` : Series id
+     * `d` : List of tracked frame types encoded as single bytes (`C`), if the list is empty
```
Comment (Irving-cl) [inline src/lib/spinel/spinel.h:3102]: When transmitting this field in Link Metrics Management Request, we use bit fields in one byte to represent this information: [SeriesFlags](https://github.com/openthread/openthread/blob/master/src/core/thread/link_metrics_tlvs.hpp#L360)  

Shall we use one byte to replace a list (data blob) here?

File: src/ncp/ncp_base_mtd.cpp [hunk 2]
```diff
@@ -2927,6 +3002,106 @@ template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_MAC_FIXED_RSS>(void)
...
+    struct otIp6Address      address;
+    uint8_t                  seriesId;
+    const uint8_t *          types;
+    uint16_t                 typesLen;
+    otLinkMetrics            linkMetrics = {};
+    otLinkMetricsSeriesFlags seriesFlags = {};
+
+    SuccessOrExit(error = mDecoder.ReadIp6Address(address));
+    SuccessOrExit(error = mDecoder.ReadUint8(seriesId));
+    SuccessOrExit(error = mDecoder.ReadDataWithLen(types, typesLen));
```
Comment (Irving-cl) [inline src/ncp/ncp_base_mtd.cpp:3073]: As mentioned above, I think we can just use one byte for SeriesFlags.

**Review by Irving-cl — APPROVED** [review]
Overall LGTM. Thanks!

A few suggestions.

**Comment by Irving-cl** [comment]
@piotrkoziar 

Please rebase the PR.

**Comment by piotrkoziar** [comment]
> @piotrkoziar
> 
> Please rebase the PR.

Rebased.

File: src/lib/spinel/spinel.h [hunk 3]
```diff
@@ -724,6 +726,43 @@ enum
...
+};
+
+// Metric ids used for:
+// @ref SPINEL_PROP_THREAD_LINK_METRICS_QUERY
+// @ref SPINEL_PROP_THREAD_LINK_METRICS_QUERY_RESULT
+// @ref SPINEL_PROP_THREAD_LINK_METRICS_MGMT_ENH_ACK
+// @ref SPINEL_PROP_THREAD_LINK_METRICS_MGMT_FORWARD
+enum
+{
+    THREAD_LINK_METRIC_PDU_COUNT   = 0,
```
Comment (LuDuda) [inline src/lib/spinel/spinel.h:750]: Add `SPINEL_` prefix.

File: src/lib/spinel/spinel.h [hunk 4]
```diff
@@ -724,6 +726,43 @@ enum
...
+    THREAD_LINK_METRIC_LQI         = 1,
+    THREAD_LINK_METRIC_LINK_MARGIN = 2,
+    THREAD_LINK_METRIC_RSSI        = 3,
+};
+
+// Frame types used for:
+// @ref SPINEL_PROP_THREAD_LINK_METRICS_MGMT_FORWARD
+enum
+{
+    THREAD_FRAME_TYPE_MLE_LINK_PROBE   = 0,
```
Comment (LuDuda) [inline src/lib/spinel/spinel.h:760]: Add `SPINEL_` prefix.

File: src/ncp/ncp_base.hpp [hunk 5]
```diff
@@ -354,6 +378,12 @@ class NcpBase
     otError EncodeChildInfo(const otChildInfo &aChildInfo);
 #endif
 
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+    otError DecodeLinkMetrics(otLinkMetrics *linkMetrics, bool allowPduCount);
```
Comment (LuDuda) [inline src/ncp/ncp_base.hpp:382]:
```suggestion
    otError DecodeLinkMetrics(otLinkMetrics *aMetrics, bool aAllowPduCount);
```

**Review by LuDuda — COMMENTED** [review]
LGTM :+1:
I would however apply @abtink suggestions regarding types.

**Comment by piotrkoziar** [comment]
@abtink @Irving-cl I made some changes as requested:

* metric ids and frame types are encoded as bit fields in single byte (like Series Flags)
* arrays with metrics from `SPINEL_PROP_THREAD_LINK_METRICS_MGMT_ENH_ACK_IE` and `SPINEL_PROP_THREAD_LINK_METRICS_QUERY_RESULT` are now embedded in struct (as @abtink proposed in https://github.com/openthread/openthread/pull/6228#discussion_r589669607)

**Comment by piotrkoziar** [comment]
@konradderda may want to look at this as well

**Comment by piotrkoziar** [comment]
rebased

File: src/ncp/ncp_base_mtd.cpp [hunk 6]
```diff
@@ -209,6 +209,81 @@ otError NcpBase::EncodeNeighborInfo(const otNeighborInfo &aNeighborInfo)
...
+    uint8_t metrics = 0;
+
+    SuccessOrExit(error = mDecoder.ReadUint8(metrics));
+
+    if (metrics & SPINEL_THREAD_LINK_METRIC_PDU_COUNT)
+    {
+        VerifyOrExit(aAllowPduCount, error = OT_ERROR_INVALID_ARGS);
+        aMetrics->mPduCount = true;
+    }
+    if (metrics & SPINEL_THREAD_LINK_METRIC_LQI)
```
Comment (abtink) [inline src/ncp/ncp_base_mtd.cpp:225]: nit: newline before `if`?
```suggestion

    if (metrics & SPINEL_THREAD_LINK_METRIC_LQI)
```

File: src/ncp/ncp_base_mtd.cpp [hunk 7]
```diff
@@ -209,6 +209,81 @@ otError NcpBase::EncodeNeighborInfo(const otNeighborInfo &aNeighborInfo)
...
+    {
+        aMetrics->mLqi = true;
+    }
+    if (metrics & SPINEL_THREAD_LINK_METRIC_LINK_MARGIN)
```
Comment (abtink) [inline src/ncp/ncp_base_mtd.cpp:229]:
```suggestion

    if (metrics & SPINEL_THREAD_LINK_METRIC_LINK_MARGIN)
```

File: src/ncp/ncp_base_mtd.cpp [hunk 8]
```diff
@@ -209,6 +209,81 @@ otError NcpBase::EncodeNeighborInfo(const otNeighborInfo &aNeighborInfo)
...
+    {
+        aMetrics->mLinkMargin = true;
+    }
+    if (metrics & SPINEL_THREAD_LINK_METRIC_RSSI)
```
Comment (abtink) [inline src/ncp/ncp_base_mtd.cpp:233]:
```suggestion

    if (metrics & SPINEL_THREAD_LINK_METRIC_RSSI)
```

File: src/ncp/ncp_base_mtd.cpp [hunk 9]
```diff
@@ -3039,6 +3114,102 @@ template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_MAC_FIXED_RSS>(void)
...
+    SuccessOrExit(error = mDecoder.ReadUint8(seriesId));
+    SuccessOrExit(error = mDecoder.ReadUint8(types));
+
+    SuccessOrExit(error = DecodeLinkMetrics(&linkMetrics, true));
+
+    if (types & SPINEL_THREAD_FRAME_TYPE_MLE_LINK_PROBE)
+    {
+        seriesFlags.mLinkProbe = true;
+    }
+    if (types & SPINEL_THREAD_FRAME_TYPE_MAC_DATA)
```
Comment (abtink) [inline src/ncp/ncp_base_mtd.cpp:3192]:
```suggestion

    if (types & SPINEL_THREAD_FRAME_TYPE_MAC_DATA)
```

File: src/ncp/ncp_base_mtd.cpp [hunk 10]
```diff
@@ -3039,6 +3114,102 @@ template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_MAC_FIXED_RSS>(void)
...
+    {
+        seriesFlags.mMacData = true;
+    }
+    if (types & SPINEL_THREAD_FRAME_TYPE_MAC_DATA_REQUEST)
```
Comment (abtink) [inline src/ncp/ncp_base_mtd.cpp:3196]:
```suggestion

    if (types & SPINEL_THREAD_FRAME_TYPE_MAC_DATA_REQUEST)
```

File: src/ncp/ncp_base_mtd.cpp [hunk 11]
```diff
@@ -3039,6 +3114,102 @@ template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_MAC_FIXED_RSS>(void)
...
+    {
+        seriesFlags.mMacDataRequest = true;
+    }
+    if (types & SPINEL_THREAD_FRAME_TYPE_MAC_ACK)
```
Comment (abtink) [inline src/ncp/ncp_base_mtd.cpp:3200]:
```suggestion

    if (types & SPINEL_THREAD_FRAME_TYPE_MAC_ACK)
```

**Review by abtink — APPROVED** [review]
@piotrkoziar, looks really great. Thank you. 👍 

Just few small style suggestions below.

**Comment by piotrkoziar** [comment]
@jwhui, I think this PR is ready. Can we merge it?

**Review by jwhui — APPROVED** [review]
Thanks! 👍🏼

**Pull request merged**